django/db/models/sql/where.py

```diff
@@ -58,8 +58,8 @@ def split_having(self, negated=False):
                 having_parts.append(c)
             else:
                 where_parts.append(c)
-        having_node = self.__class__(having_parts, self.connector, self.negated) if having_parts else None
-        where_node = self.__class__(where_parts, self.connector, self.negated) if where_parts else None
+        having_node = self.create(having_parts, self.connector, self.negated) if having_parts else None
+        where_node = self.create(where_parts, self.connector, self.negated) if where_parts else None
```

Comment:
So I'm going through your changes and figuring out which bits to pull back in as per Mariusz' request. This one confuses me, as I'm intending to pull 2399d5da1d2f842be45fee17dacb52e68592737f which though I've not benchmarked it, probably makes sense. It sounds correct, anyway:

> As `WhereNode` doesn't override `__init__()` we can just use `self.__class__()` directly which is faster. As an aside, we also do not need to pass in `children=[]`. That is handled by `Node.__init__()`

But here you've gone back to using `create`/`_new_instance`. Is that because the `__class__` can end up as a `Q` rather than as I'd guess, a `WhereNode`?

Reply:
Heya. Just seen this. Will reply later this evening. 🙂

Reply:
Hey @kezabelle. So I mentioned about bringing `._new_instance()` back in https://github.com/django/django/pull/14673#discussion_r673884166 but I wanted to "bless" it more by renaming it to `.create()` (while still keeping it private).

This isn't really going to be ready for the feature freeze, so let's not rush it. I'm hoping to give it some attention over the next couple of weeks. My thoughts for this were:

- Pick "Remove obsolete docstring from `WhereNode.clone()`."
- Bring in your changes from #14709.
- Rename `Node._new_instance()` to `Node.create()`.
- Make use of `.create()` to replace calls using `type(self)(...)`, `self.__class__(...)`, `self.__class__.create(...)`, and `Node(...)`.
- I also figured we could use it to follow up on #14699 and switch (again) to use `.create()` to avoid `*`-unpacking and the `super().__init__()` call in `Q.__init__()`. It'll make most of the construction of these `Node` (sub)classes consistent.
- And then, as per the end of the description on ticket-32970 we implement `__copy__` and `__deepcopy__` with `.copy()` and `.clone()` as aliases to `__copy__`. We put this on `Node` as all subclasses are pretty much doing the same thing. (The aliases are just for backward compatibility, for now at least).
- And then there is just the tidying of the messy if-statements, etc.

Just to note, because of the way `._new_instance()/.create()` is implemented, I expect that we just end up with:

```python
class Node:
    def __init__(self, children=None, connector=None, negated=False):
        self.children = children[:] if children else []
        self.connector = connector or self.default
        self.negated = negated

    @classmethod
    def create(cls, children=None, connector=None, negated=False):
        obj = Node(children, connector, negated)
        obj.__class__ = cls
        return obj

    def __copy__(self):
        return self.create(children=self.children, connector=self.connector, negated=self.negated)

    copy = __copy__
    clone = __copy__

    def __deepcopy__(self, memodict):
        obj = self.create(connector=self.connector, negated=self.negated)
        obj.children = copy.deepcopy(self.children, memodict)
        return obj
```

Reply:
I'm glad I asked. I sincerely appreciate the amount of thought you've put into it, compared to my piecemeal scattergun approach across multiple PRs :)

Reply:
@kezabelle So I've quickly hacked together what I think this ought to look like. The intent was to try and encapsulate the funny behaviour of setting `__class__` inside `Node.create()` and then use that everywhere.

(After playing with this, a further thing to do that would be nice is to unify `.copy()` and `.clone()` if possible.)